tensorflow/compiler/jit/extract_outside_compilation_pass.cc

```diff
@@ -1240,11 +1241,10 @@ Status RewriteShapeInferenceGraph(const string& shape_inference_graph_name,
     // This is an "top-level" outside compilation. Clear the graph, and copy
     // SendFromHost and all its predecessors from `host_graph`.
     std::vector<Node*> nodes;
+    nodes.reserve(g->num_op_nodes());
     for (Node* n : g->op_nodes()) {
```

Comment:
These two loops should not be merged... We generally don't mutate vector when iterating on it.

Reply:
But wouldn't this be more efficient, even considering the cache?

Reply:
Does your change (using the same loop for iteration & removing node) work now?
Yes.

Is it guaranteed to work if tensorflow::Graph::op_nodes(), tensorflow::Graph::RemoveNode() and tensorflow::NodeIter changes their implementations?
No. And this is why I am suggesting to use two loop instead.

-- TLDR --

Today your change works because **we never really remove elements from op_nodes vector**. When we call tensorflow::Graph::RemoveNode(), we just set that node in the vector to nullptr. And when tensorflow::NodeIter iterates through the op_nodes vector, it skips those (removed) nullptr nodes so we only get valid nodes. As a result, iterator for op_nodes is still valid after removing elements from op_nodes, and you can remove nodes from op_nodes vector and continue iterating without any issues.

But the above assumption is not part of the tensorflow::Graph interface, so it might change. Imagine one day, for some reason, tensorflow::Graph::op_nodes() is changed to std::set<Node*>, and tensorflow::Graph::RemoveNode(Node*) really removes a node from that set. Then iterator for op_nodes will become invalid after removing elements from op_nodes, so we cannot remove elements while iterating on op_nodes.

Reply:
Good point. Changed.